Enhancement/update success error messages to wp admin notices

MaxwellGarceau opened the pull request:
<!--
Filling out this template is required.  Any PR that does not include enough information to be reviewed may be closed at a maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

## Description of the Change
<!--
We must be able to understand the design of your change from this description.  The maintainer reviewing this PR may not have worked with this code recently, so please provide as much detail as possible.

Where possible, please also include:
- verification steps to ensure your change has the desired effects and has not introduced any regressions
- any benefits that will be realized
- any alternative implementations or possible drawbacks that you considered
- screenshots or screencasts
-->

On the plugin settings page, replace the custom Mailchimp UI for admin notices to the standard WP admin notices API

![mailchimp-messages-to-wp-admin-notices](https://github.com/user-attachments/assets/53c56c99-44e5-4fd4-8fd3-90135cdf4c96)

### Drawbacks

- Two systems for displaying messages. WP admin notices and the Mailchimp custom messages for the front end.

### Possible regressions
While I've tested for all of these scenarios and fixed any issues I've found the following items are the most likely regressions.

- It's possible that some messages that should display as WP admin notices were missed and will still display as custom Mailchimp messages
  - It's possible that those messages that were not updated will not display at all due to removing custom message output on the plugin settings page
- It's possible that specialized message text will be stripped out by sanitization functions that have been introduced or tweaked

<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
Closes #62 

## How to test the Change
<!-- Please provide steps on how to test or validate that the change in this PR works as described. -->

### WP Admin messages
A video recording can be made on request for any of the testing items below.

I've documented all of the possible testing scenarios that I found while making the changes, but due to the extensive list of areas to test I recommend making a note of any items that are too difficult or time consuming to efficiently test. We can then look at the items that were not tested and decide on a plan.

For example, testing the error message for an invalid Mailchimp API token would require finding a way to fail the OAuth flow while still navigating to the plugin settings page. It might be better to talk about items like that and brainstorm easier ways of ensuring regression proof code, such as relying on the OAuth test in `/tests/cypress/e2e/connect.test.js` or extending it.

![image](https://github.com/user-attachments/assets/f28c6742-7b90-4c69-bccb-18d19aaa8146)

#### Success message

Settings page URL: `/wp-admin/admin.php?page=mailchimp_sf_options`

- Save any option and the success message will display
  - (optional) If we really want to go above and beyond we could save every single option on the page to trigger all of the locations where the success message will display
  - Pro tip: Multiple admin notices will display at the same time. You can turn every option on and then every option off again. Caution should be taken for options that will conditionally hide or display other options.

#### Error message

- Set the users list to "-- Select a List --" and click "Update List"
- If a users lists can not be retrieved from Mailchimp then an error will display
- If a user does not have any lists in their Mailchimp account defined then an error will display

### FE messages
**FE messages should remain unchanged**

![image](https://github.com/user-attachments/assets/960812ef-81a1-4bd7-899f-2f16403d3a25)

#### Success Messages
Submit the form after filling out all required fields to trigger the success message

- Users that already exist as a contact will be subscribed without email verification required (I think)
- Users not already saved as a contact will receive a success message to verify their email

#### Error Messages
Submit the form without entering any details to trigger the error message
The following scenarios will also throw errors on the FE

- Required fields not filled out or invalid
- If Mailchimp API client won't work (invalid token, etc) then an error will be thrown
- If update existing is turned off in the plugin settings and the subscriber email already exists then an error will be thrown
- Error with API request adding subscriber to list (could trigger with chrome dev tools -> mobile device view -> go offline)

## Changelog Entry
<!--
Please include a summary for this PR, noting whether this is something being Added / Changed / Deprecated / Removed / Fixed / or Security related.  You can replace the sample entries after this comment block with the single changelog entry line for this PR. -->
> Changed - Plugin settings page success and error messages will now use WP admin notices

## Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @MaxwellGarceau 

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/mailchimp/wordpress/blob/develop/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

MaxwellGarceau commented on includes/admin/admin-notices.php:
```diff
@@ -0,0 +1,78 @@
+<?php
```
I added a new file to modularize admin notice helper functions and I'm requiring it inside `includes/class-mailchimp-admin.php`.

Is this okay or is another project structure preferred?

dkotter replied:
Location of this is fine by me. We could add a namespace to this file and remove the `mailchimp_sf_` prefix from the functions, which is something we typically do on projects though hasn't been followed yet within this plugin

MaxwellGarceau replied:
Good idea, @dkotter. I've added the namespace to the `admin-notices.php` file functions and removed the `mailchimp_sf_` prefix in commit 4fda85e.

MaxwellGarceau commented on includes/class-mailchimp-admin.php:
```diff
@@ -38,6 +38,26 @@ public function init() {
 		add_action( 'admin_enqueue_scripts', array( $this, 'enqueue_admin_page_scripts' ) );
 		add_action( 'admin_menu', array( $this, 'add_create_account_page' ) );
 		add_filter( 'admin_footer_text', array( $this, 'admin_footer_text' ) );
+
+		// TODO: Should this load somewhere else?
+		$this->require_admin_utils();
+	}
+
+	/**
+	 * Require admin utils.
+	 *
+	 * Currently includes admin notices from a single file. Long term plan is to break up admin
+	 * functionality into smaller, more focused files to improve maintainability. This could also include:
+	 * - Moving OAuth related code to oauth.php
+	 * - Moving account creation code to account.php
+	 * - Moving settings page code to settings.php
+	 * - Moving notices code to notices.php (already done)
+	 * This will help avoid having too much code in a single file and make the codebase more modular.
+	 *
+	 * @since 1.7.0
+	 */
+	private function require_admin_utils() {
+		include_once MCSF_DIR . 'includes/admin/admin-notices.php';
```
This is where I'm requiring `includes/admin/admin-notices.php`. I added some notes regarding other code that could be modularized over time.

What does everyone think?

dkotter replied:
That list looks good to me and I'm always open to better code organization

MaxwellGarceau commented on mailchimp.php:
```diff
@@ -253,7 +253,7 @@ function mailchimp_sf_request_handler() {
 							header( 'Last-Modified: ' . gmdate( 'D, d M Y H:i:s' ) . ' GMT', true, 200 );
 						}
-						echo wp_kses_post( mailchimp_sf_global_msg() );
+						echo wp_kses_post( mailchimp_sf_frontend_msg() );
```
Renamed `mailchimp_sf_global_msg` to `mailchimp_sf_frontend_msg` to reflect new role of this msg function.

Flagging for extra scrutiny.

dkotter replied:
I'm fine with the rename though I guess not sure if it's accurate? Seems that function still outputs admin messages, while I'd expect a function called frontend message to only output messages on the front-end, not admin?

MaxwellGarceau replied:
Hey @dkotter, thanks for calling this out. Are you referring to the `mailchimp_sf_frontend_msg` call in `mailchimp_sf_request_handler` at line 256? I believe this fetches the error messages accumulated during the BE validation handling.

There's some validation on the accounts creation page in the admin, but that looks like it's being handled via JS validation in `assets/js/admin.js`.

I double checked the codebase for other areas where `mailchimp_sf_request_handler` might be used in the admin, but I couldn't find any. Please correct me if I'm wrong. I'd be happy to revert the same or update the codebase.

dkotter replied:
Ah, okay, it appeared this function was using the `admin_notice_success` method, so I assumed it was outputting admin messages. But that was my mistake in not looking close enough at the diff here. This all looks good to me

MaxwellGarceau commented on mailchimp.php:
```diff
@@ -296,7 +296,7 @@ function mailchimp_sf_migrate_sopresto() {
 		} catch ( Exception $e ) {
 			$msg = '<strong class="mc_error_msg">' . $e->getMessage() . '</strong>';
-			mailchimp_sf_global_msg( $msg );
+			mailchimp_sf_frontend_msg( $msg );
```
I couldn't find any references to `mailchimp_sf_migrate_sopresto` in the codebase.

Can we remove this function? If not, can we mark it as deprecated?

dkotter replied:
I'd have to look through the code history to be sure but I think this was an old feature that is no longer used. I wouldn't worry about removing it here but worth opening a new Issue to track that removal

MaxwellGarceau replied:
Moved to issue #94.

MaxwellGarceau reviewed (commented):
I've highlighted some areas that I think would benefit from additional scrutiny. If anyone would like to request changes I would be happy to make them =).

MaxwellGarceau commented:
@dkotter @jeffpaul currently admin notice messages will display in the admin and the plugins own message system for displaying messages on the front end.

Do you think it's worth it to add a section in the readme regarding this pattern for documentation or do you think it's overkill?

dkotter commented on includes/class-mailchimp-admin.php:
```diff
@@ -38,6 +38,26 @@ public function init() {
 		add_action( 'admin_menu', array( $this, 'add_create_account_page' ) );
 		add_filter( 'admin_footer_text', array( $this, 'admin_footer_text' ) );
+
+		// TODO: Should this load somewhere else?
```
I think this is a fine location for this so this `TODO` comment can be removed

MaxwellGarceau replied:
Removed

jeffpaul commented:
> Do you think it's worth it to add a section in the readme regarding this pattern for documentation or do you think it's overkill?

I don't think this needs to be called out in the readme.

MaxwellGarceau commented on .github/workflows/build-release-zip.yml:
```diff
@@ -35,6 +35,11 @@ jobs:
         if: steps.cache-node-modules.outputs.cache-hit != 'true'
         run: npm ci --no-optional
 
+      - name: Install Composer dependencies and dump autoload
+        run: |
+          composer install --no-dev --optimize-autoloader
+          composer dump-autoload
+
```
Install composer deps

MaxwellGarceau commented on .github/workflows/wordpress-plugin-deploy.yml:
```diff
@@ -19,6 +19,11 @@ jobs:
         node-version-file: '.nvmrc'
         cache: 'npm'
 
+    - name: Install Composer dependencies and dump autoload
+      run: |
+        composer install --no-dev --optimize-autoloader
+        composer dump-autoload
+
```
Install composer deps

MaxwellGarceau commented on composer.json:
```diff
@@ -18,6 +18,17 @@
     "10up/phpcs-composer": "^3.0",
 	  "sirbrillig/phpcs-changed": "^2.11"
   },
+  "autoload": {
+    "psr-4": {
+      "Mailchimp\\WordPress\\": "src/"
+    },
+    "classmap": [
+      "includes"
+    ],
+    "files": [
+      "includes/admin/admin-notices.php"
+    ]
+  },
```
Autoload

MaxwellGarceau commented on mailchimp.php:
```diff
@@ -34,6 +34,19 @@
 
+use function Mailchimp\WordPress\Includes\Admin\{admin_notice_error, admin_notice_success};
+
+// Define the path to the Composer autoload file
+$autoload_file = __DIR__ . '/vendor/autoload.php';
+
+// Check if the autoload file exists
+if ( ! file_exists( $autoload_file ) ) {
+	die( 'Composer autoload file not found. Run `composer install` to generate it.' );
+}
+
+// Include the autoload file
+require_once $autoload_file;
```
Autoload

MaxwellGarceau reviewed (commented):
@dkotter, I also added autoloading to have an easy way to modularize new code and refactor existing code in the future.

**❗I modified the Github action workflow files to install composer dependencies in order to ensure autoload was being generated on every build. I wanted to flag this in case I missed a required build process update.**

What do you think? I can revert the commit if you think it's better to keep the existing `require` strategy or to tackle in a separate issue.

Also, if there's a good place to add documentation for in progress pattern updates I'd be happy to add switching to autoload in new work and any refactors.

MaxwellGarceau commented:
I removed the composer autoload and changes to the build process to keep this PR modular.

Composer autoload discussion continued on issue #96 and PR #95.

qasumitbagthariya commented:
### QA Update  ✅ 

-------------

I have verified this PR in the `enhancement/update-success-error-messages-to-wp-admin-notices` branch which has been fixed and is functioning as intended.

I tested the following on this branch:

**Update list**
<img width="1499" alt="image" src="https://github.com/user-attachments/assets/e56f3336-06f2-489e-a0fa-89924214a150" />


**Update form setting**
<img width="1503" alt="image" src="https://github.com/user-attachments/assets/98de5541-c22c-4e59-8422-7d793caac9cd" />

**no valid list**
<img width="1504" alt="image" src="https://github.com/user-attachments/assets/ca3801d6-922e-45c3-85de-ea0bfd8ef052" />

**Connection problme error**
<img width="1508" alt="image" src="https://github.com/user-attachments/assets/c1f4ed56-c885-4b94-b9bb-a7637cc266de" />

**Testing Environment** 

<Details>

- WordPress: 6.7.1
- Theme: Twenty Twenty-Four 1.3
- WooCommerce - 9.5.1
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS 15.2 
- Branch: enhancement/update-success-error-messages-to-wp-admin-notices

</Details>


Steps to Test- As mentioned in the PR description.
Test Results - It is working as expected.
Functional Demo / Screencast - 
Special Notes -  Ready for code review (Woo)
Testing Document status:
Cases related to this Issue/PR are added to the Critical Flow Wiki pages: 

- [ ] Yes
- [x] Not Required/Applicable for this PR
------------------------------------------------	

**Question:**

Why do we display the following messages every time a setting is saved?  
- Fancy JavaScript submission turned ON/OFF!  
- Datepicker turned ON/OFF!  
- Double opt-in turned ON/OFF!

<img width="1511" alt="image" src="https://github.com/user-attachments/assets/e66cd59a-531f-4168-b804-a537df61bb34" />

MaxwellGarceau commented:
@vikrampm1 do you want me to rebase these changes with develop? It might be hard for another dev to accurately fix the merge conflict and I don't mind, but checking in first.

vikrampm1 commented:
@MaxwellGarceau sure, if it doesn't take too much of your time then please go ahead. Otherwise, please have @iamdharmesh help and progress this PR further.

MaxwellGarceau commented:
> @MaxwellGarceau thanks for the PR! Could you please rebase your PR on top of the latest changes in the base branch?

### Rebased changes
Rebased `enhancement/update-success-error-messages-to-wp-admin-notices` with changes from `develop`.

Used rebase specifically because I was prompted by the PR bot, but let me know if another method is preferred in the future. A backup of the original branch can be found at [before-rebase/enhancement/update-success-error-messages-to-wp-admin-notices](https://github.com/mailchimp/wordpress/tree/before-rebase/enhancement/update-success-error-messages-to-wp-admin-notices)

Fixed merge conflicts and verified changes by testing validation and submission locally with the GUI.

### Autoload
Added autoloading to the admin notice functions that was not possible at the original time of development due to the composer autoloading changes not being merged to develop.

qasumitbagthariya commented:
@iamdharmesh 

**Note:**

In the success message displayed after saving the Mailchimp list data, there is no space between the list name "mailchimp" 

![image](https://github.com/user-attachments/assets/4f9cd635-fdd6-417f-a254-9fbc2e15f314)

iamdharmesh commented:
> In the success message displayed after saving the Mailchimp list data, there is no space between the list name "mailchimp"

Fixed in https://github.com/mailchimp/wordpress/pull/85/commits/e9b64650ceadc3121d5a77a43d829164e705e4d7

qasumitbagthariya commented:
Thanks for the fix @iamdharmesh 

Regression / Smoke Test Report  ✅ 
-------

Tested with the `smoke-testing` branch, it works as expected, similar to the fix-specific branch.

<img width="1504" alt="image" src="https://github.com/user-attachments/assets/8ba6019e-28aa-41cf-ac4d-e3c4c92cf95b" />

**Testing Environment** 

<Details>

- WordPress: 6.7.2
- Theme: Storefront 4.6.0
- Theme: Twenty Twenty-Four 1.3
- PHP: 8.2.23
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS 15.2 
- Branch: smoke-testing

</Details>


Next Step- Ready to Merge 🚀
